[core][sql] enum SPELLTYPE and spell_list.sql enum Implementation

DotesMiqotes opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

<!-- Describe what your PR does here. If it closes an existing issue, you can mention: "Closes #1234" and GitHub will link this PR to that issue. -->
Implements `enum SPELLTYPE` to clean up and optimize the `bool is()` functions in `spell.cpp`.
Implements enum style variables in `spell_list.sql` for the `group` `family` `element` `validTargets` `skill` and `type` columns.
Adds a `type` column to `spell_list.sql`.
Adds `setSpellType()` to `spell.cpp` to load the spell types from the database.
Adds `getSpellType()` to `spell.cpp` that returns the spell type flags.
Aranges the columns of `spell_list.sql` into a more readable format.
Fixes every instance of a spell having the wrong element, valid targets, or skill.
Adds all necessary types to all ninjutsu spells and bard songs to ensure they sort properly into their respective mob spell containers.

## Steps to test these changes

<!-- Clear and detailed steps to test your changes here -->
I compiled the server, and used `source sql/spell_list.sql` to update the database. I then play tested the changes by running around and fighting as many types of casters I could in Castle O with a party full of trusts. Trusts, mobs I was fighting, idle mobs, and their pets were all casting everything they should have been as expected.
I'm using the same testing method after upgrading `spell_list.sql` and everything is working as expected. Mobs are buffing themselves, nuking and debuffing me as well as my trusts when I engage them. My trusts are all casting the spells they should be, and I'm having no issues with any of my spells on RNG/RDM.

DotesMiqotes committed:
[core] enum SPELLTYPE Implementation

Implements `enum SPELLTYPE` to clean up and optimize the `bool is()` functions in `spell.cpp`.
Adds a `type` column to `spell_list.sql`.
Adds `setSpellType()` to `spell.cpp` to load the spell types from the database.
Adds `getSpellType()` to `spell.cpp` that returns the spell type flags.

zach2good reviewed (commented):
Can spells have multiple `SpellTypes`? Otherwise, there's no need to store/use this as a flag and check everything as `spellType & SPELLTYPE_HEAL`. I'd argue it's cleaner and easier to understand in the `spell->isHeal()` form.

Does the current implementation cover all spells of all types?

Just as a point of styling that isn't written down anywhere: it doesn't feel so nice to put this new numeric type column in the table at the end of the row, where a few other tables keep the last entry for the expansion string. It also winds up being another magic number that you have to go back and forth to know/remember what it is. I think you can see in the items tables we've started using aliases to make these sort of things more human readable.

DotesMiqotes commented:
Under the current logic, they're calculated as if they have multiple types.

This covers all spells in the database, I went through line by line making sure they had every type the server was calculating them as. I also ninja fixed the skill of Gravity II (it wasn't enfeebling).

I debated putting the `type` column either just before or after the `skill` column, but it would have made it even less readable than it is now. I felt that `content_tag` acts as a nice reference point in the row to count out from. It definitely made things a bit easier on me when I was determining types. Moving it later is a fairly easy task if needed.

YAML databases would also be great for static stuff like this. They're meant to be human readable. From what I remember working with them on rAthena, they were a bit more efficient than SQL as well.

DotesMiqotes commented:
Quick update, I'm going to be converting more than the new magic numbers to aliases.

Xaver-DaRed commented:
Being honest, theres sonething about this that just feels wrong to me.
For starters the enums themselves. Spells could fall into many categories, sure, for instance, healing spells CAN be damage spells.

However, this enum list doesnt have a damage category, nor an aborb category (drain, aspir, etc) nor many others.
Probably for good reaso: they arent used nor needed in the code.
So that leads me to believe that this categories are just minor exceptions, and dont really require this rework.
And that leads to another thing: The database.

I can only speak for myself, but, I personally want to see less static data in the database, not more.
The reason for this is that, the more data theres in the database, the less readable it is and the less likely ppl areto come to help fix it.

In short: My real question here would be "Why is this being done?" or "Whats the issue we are trying to fix with this?"

No offense intended, just my opinion and doubts on the matter.

DotesMiqotes commented:
The logic is predetermined and types don't need to be recalculated every time a spell is cast, or added to a spell list container. This change collapses multiple if statements into one every time the spell type flag is used. I didn't add anything for damage or drains because in the cases a type is used for them, they're sorted into a catch all. My other ninja fix involved that, the logic determined that all debuffs were being sorted into a mob's damage spell container, and the debuff container was staying empty. Those 'special cases' are most of `spell_list.sql`.

I already have the entire type column converted to aliases now as well, it's done wonders for readability. I plan on doing the same with a few of the other columns before I make the commit.

This was one of the improvements that I found I could make when I was doing the research needed to address this:
https://github.com/LandSandBoat/server/blob/be17f7f3c3edfe40e12576854da0a17c2bf2c481/scripts/actions/spells/trust/joachim.lua#L30

DotesMiqotes committed:
[sql] Implements enums in spell_list.sql

Implements enum style variables in `spell_list.sql` for the `group` `family` `element` `validTargets` `skill` and `type` columns.
Aranges the columns into a more readable format.
Fixes every instance of a spell having the wrong element, valid targets, or skill.
Adds all necessary types to all ninjutsu spells and bard songs to ensure they sort properly into their respective mob spell containers.

Xaver-DaRed commented on src/map/spell.h:
```diff
@@ -234,6 +234,18 @@ enum SPELLFLAG
+    SPELLTYPE_NA     = 0x10,
+    SPELLTYPE_RAISE  = 0x20,
+    SPELLTYPE_SEVERE = 0x40,
+};
```
Whats the difference between heal and cure?

Xaver-DaRed replied:
Heal is healing magic, in general, ok

Cure is cure and cura, families... Why not use families then?

DotesMiqotes replied:
Types are based on this. On their own `SPELLTYPE_CURE` and `SPELLTYPE_HEAL` won't get used much if ever, Valid targets are taken into consideration when determining the actual spell type. Cure, Curaga, and Cura also 3 separate families.
```c++
bool CSpell::isHeal()
{
    return ((getValidTarget() & TARGET_SELF) && getSkillType() == SKILL_HEALING_MAGIC) || m_ID == SpellID::Pollen || m_ID == SpellID::Wild_Carrot ||
           m_ID == SpellID::Healing_Breeze || m_ID == SpellID::Magic_Fruit;
}
```
```c++
bool CSpell::isCure()
{
    return ((static_cast<uint16>(m_ID) >= 1 && static_cast<uint16>(m_ID) <= 11) || m_ID == SpellID::Cura || m_ID == SpellID::Cura_II ||
            m_ID == SpellID::Cura_III);
}
```
```c++
bool CSpell::isBuff() const
{
    return (getValidTarget() & TARGET_SELF) && !(getValidTarget() & TARGET_ENEMY);
}
```
```
Cure   = @G_WHITE,@F_CURE  ,@E_LIGHT,@VT_CURESPL ,@S_HEALING,@T_HMCURE
Curaga = @G_WHITE,@F_CURAGA,@E_LIGHT,@VT_SLF_PRTY,@S_HEALING,@T_BUFFHMCURE
Cura   = @G_WHITE,@F_CURA  ,@E_LIGHT,@VT_SELF    ,@S_HEALING,@T_BUFFHMCURE
```

Xaver-DaRed commented on src/map/spell.cpp:
```diff
@@ -452,7 +424,7 @@ namespace spell
     {
         const char* Query = "SELECT spellid, name, jobs, `group`, family, validTargets, skill, castTime, recastTime, animation, animationTime, mpCost, \
-                             AOE, base, element, zonemisc, multiplier, message, magicBurstMessage, CE, VE, requirements, content_tag, spell_range \
+                             AOE, base, element, zonemisc, multiplier, message, magicBurstMessage, CE, VE, requirements, content_tag, type, spell_range \
```
While we are here, why dont we order this properly, following the same order as the database?

DotesMiqotes replied:
Sounds good. I'll make it happen.

DotesMiqotes replied:
```sql
INSERT INTO `spell_list` VALUES (1,'cure',@G_WHITE,@F_CURE,@E_LIGHT,@VT_CURESPL,@S_HEALING,@T_HMCURE,0,8,2000,5000,7,7,1,2000,4,0,1.00,0,0,0,204,@CT_BASE,'WHM:1,RDM:3,PLD:5,SCH:5');
```

DotesMiqotes replied:
RIP `requirements`
```sql
INSERT INTO `spell_list` VALUES (16,'blindna',@G_WHITE,@F_NA,@E_LIGHT,@VT_SUPPSPL,@S_HEALING,@T_HMNA,0,16,1000,10000,0,0,16,2000,4,0,1.00,1,165,4,204,@CT_BASE,'WHM:14,SCH:17:AW');
```

Xaver-DaRed replied:
I will say, if this grows too big, it will be a nightmare to review properly.

Maybe the simpler changes should go in separate PRs first.

DotesMiqotes replied:
There won't be anything difficult to review by the time I'm done. I already reordered everything in the SQL query, and arranged everything that comes after before starting this change. It won't take me much to rework the logic required to read the database.
```sql
INSERT INTO `spell_list` VALUES (11,'curaga_v',@G_WHITE,@F_CURAGA,@E_LIGHT,@VT_SLF_PRTY,@S_HEALING,@T_BUFFHMCURE,0,380,5750,11000,0,0,11,2000,1,0,1.00,0,0,204,@CT_ABY,'WHM:91');
INSERT INTO `spell_list` VALUES (12,'raise',@G_WHITE,@F_RAISE,@E_LIGHT,@VT_PDEAD,@S_HEALING,@T_RAISE,0,150,15000,60000,0,0,12,2000,0,0,1.00,0,0,204,@CT_BASE,'WHM:25,RDM:38,PLD:50,SCH:35');
INSERT INTO `spell_list` VALUES (13,'raise_ii',@G_WHITE,@F_RAISE,@E_LIGHT,@VT_PDEAD,@S_HEALING,@T_RAISE,0,150,14000,60000,0,0,13,2000,0,0,1.00,0,0,204,@CT_BASE,'WHM:56,RDM:95,SCH:70:AW'); -- req 4
```

DotesMiqotes replied:
If it makes you feel any better about the scope of this, I'm pulling all the job and level data off of the scrolls at the auction house to put in the new col.
```sql
INSERT INTO `spell_list` VALUES (152,'blizzard_iv',@G_BLACK,@F_BLIZZARD,@E_ICE,@VT_ENEMY,@S_ELEMENTAL,@T_NONE,0,162,6000,30000,2,252,152,2000,0,506,2.00,0,0,204,@CT_BASE,'BLM:74,RDM:89,SCH:74:AB,GEO:88'); -- req 2
INSERT INTO `spell_list` VALUES (153,'blizzard_v',@G_BLACK,@F_BLIZZARD,@E_ICE,@VT_ENEMY,@S_ELEMENTAL,@T_NONE,0,267,10000,45000,2,252,153,2000,0,829,2.30,0,0,204,@CT_ABY,'BLM:89,SCH:95:AB'); -- req 2
INSERT INTO `spell_list` VALUES (154,'aero',@G_BLACK,@F_AERO,@E_WIND,@VT_ENEMY,@S_ELEMENTAL,@T_NONE,0,6,500,2000,2,252,154,2000,0,25,1.00,0,0,204,@CT_BASE,'BLM:9,RDM:14,DRK:17,SCH:12,GEO:14');
```
for example
req 1 looks like merit ability from what I can see, could easily be "WHM:75:M"
req 2 is Addendum: Black "SCH:99:AB"
req 4 is Addendum: White "SCH:99:AW"
req 8 is Tabula Rasa, and I have that for Embrava now "SCH:5:TR"
req 16 is for the latest blu stuff
req 32 is for geo stuff

DotesMiqotes commented:
Still a WIP but I wanted to give an update on where I'm at.
```sql
CREATE TABLE `spell_list` (
  `id` smallint(3) unsigned NOT NULL,
  `name` varchar(24) NOT NULL,
  `group` tinyint(3) unsigned NOT NULL DEFAULT '0',
  `family` smallint(4) unsigned NOT NULL DEFAULT '0',
  `element` tinyint(3) unsigned NOT NULL DEFAULT '0',
  `validTargets` smallint(3) unsigned NOT NULL DEFAULT '0',
  `skill` tinyint(2) unsigned NOT NULL DEFAULT '0',
  `type` tinyint(2) unsigned NOT NULL DEFAULT '0',
  `mpCost` smallint(4) unsigned NOT NULL DEFAULT '0',
  `baseDMG` smallint(4) unsigned NOT NULL DEFAULT '0',
  `multiplier` float(4,2) NOT NULL DEFAULT '1.00',
  `CE` int(10) unsigned NOT NULL DEFAULT '0',
  `VE` int(10) unsigned NOT NULL DEFAULT '0',
  `castRange` smallint(3) unsigned NOT NULL DEFAULT '0',
  `castTime` smallint(5) unsigned NOT NULL DEFAULT '0', -- TOENUM
  `recastTime` int(10) unsigned NOT NULL DEFAULT '0', -- TOENUM
  `message` smallint(5) unsigned NOT NULL DEFAULT '0', -- IN PROCESS ENUM
  `magicBurstMessage` smallint(5) NOT NULL DEFAULT '0', -- TOENUM
  `animation` smallint(5) unsigned NOT NULL DEFAULT '0',
  `animationTime` smallint(4) unsigned NOT NULL DEFAULT '2000',
  `AOE` tinyint(1) unsigned NOT NULL DEFAULT '0', -- TOENUM
  `content_tag` tinyint(3) unsigned NOT NULL DEFAULT '0',
  `requirements` varchar(50) NOT NULL DEFAULT '',
  PRIMARY KEY (`id`)
)
```
```c++

    uint16     getID() const;
//--std:string getName() const;
    uint8      getGroup() const;
    uint16     getFamily() const;
    uint16     getElement() const;
    uint16     getValidTargets() const;
    uint8      getSkill() const;
    uint8      getType() const;
    uint16     getMPCost() const;
    uint16     getBaseDMG() const;
    float      getMultiplier() const;
    uint16     getCE() const;
    uint16     getVE() const;
    float      getCastRange() const;
    uint32     getCastTime() const;
    uint32     getRecastTime() const;
    uint16     getMessage() const;
    uint16     getMagicBurstMessage() const;
    uint16     getAnimation() const;
    uint16     getAnimationTime() const;
    uint8      getAOE() const;
    uint8      getContentTag() const;

    uint8      getJob(JOBTYPE JobID) const; // TODO
    uint8      getRequirements() const;     // TODO
    uint16     getZoneMisc() const;         // TODO
```
Mostly relaxing and reorganizing things right now, I'll update with a few lines of the DB after I reorder them.

DotesMiqotes commented:
Please merge my PRs.

DotesMiqotes commented:
Good luck with things.

zach2good commented:
As a note for your future endeavours with open-source software, crowd-sourced development, and working with distributed groups of people: "Please merge my PRs" will not make maintainers swarm to review and merge your PRs.

All the best out there.
